### PR TITLE
Resolve suppressed typing warnings in filescanner.py

### PR DESCRIPTION
**Pull request by c0llab0rat0r**
`filescanner.py` had a number of type validation warnings suppressed. Most were in these categories:

* Use of generic on a named tuple `FSNodeEntry`
* Type variable not supplied on usage of the type of the named tuple
* `mypy` getting confused by usage of `Optional` with a complex recursive type
* Incorrect assumptions about standard api return types, such as from `os.fsencode` 
* A complex union of two iterators of tuples, one with three elements and the other with four, and an incorrect assumption about an `Iterator` providing a `close()` method when the api does not publish that method
* `PyCharm` disliking using `typing.AnyStr` in spite of the `typing` library explicitly having a comment on its declaration of `typing.AnyStr` that it is a public api of the `typing` module. Since this thoroughly polluted the IDE with errors, `filescanner` now has its own definition of `AnyStr` consistent with how other type variables would typically be declared

There's still some more work to do, but this hit a lot of it, and is enough for this pr.

### CODE REVIEW

**Comment by c0llab0rat0r** [comment]
@ntninja this is ready to look at. I recommend looking at the commits individually first to see the flow of the changes.

**Review by ntninja — CHANGES_REQUESTED** [review]
Just some minor stuff and questions, also the renaming `ty.AnyStr` to `AnyStr` is something I'd like to avoid.

File: tox.ini
Comment (c0llab0rat0r) [reply tox.ini:24]: `py-cid` is a transitive dependency of `pytest-cid` and was removed to deduplicate the code. It's still a dependency - now indirect.

I can add it back if you like.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:13]: Yes.

It's impossible to get PyCharm to accept the current usage. I looked for a way, then did the refactor. After the refactor I came across the comment next to AnyStr in the typing code base, and then looked again for a way to get PyCharm to accept it. It won't.

Given how broadly used PyCharm is and that this only adds a single line of code (it actually reduces the code through the rest of the file), we should keep this.

It's far too difficult to try to resolve typing errors without the IDE's help, and it gives up when it sees the reference to an outside `TypeVar`.

File: tox.ini
Comment (c0llab0rat0r) [reply tox.ini:161]: That's useful to know.

I'm going to carve this out into a separate PR.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:634]: Done. A previous flavor of this method (not committed) had been using iterators.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:658]: It's a `PyCharm` IDE warning suppression at the statement level.

File: ipfshttpclient/filescanner_ty.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner_ty.py:8]: I discovered that the problems were actually broader than `mypy`

I don't recall the specific errors I was getting, but the `attr` library has trouble with generic tuples, and so did another library.

Here's a link relevant to Python 3.6 itself:

https://stackoverflow.com/questions/50530959/generic-namedtuple-in-python-3-6

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:11]: Sounds good.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:463]: My inclination is to communicate that `.send` is out of play by using `None` but if you'd prefer they all be `Any` instead I'll go with it.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:693]: Neither `os.walk` nor `os.fwalk` declare a return type that has a `close` method.

Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:693]: Here's the implementation of `os.fwalk`:

    def fwalk(top=".", topdown=True, onerror=None, *, follow_symlinks=False, dir_fd=None):
        """Directory tree generator.

        This behaves exactly like walk(), except that it yields a 4-tuple

            dirpath, dirnames, filenames, dirfd

        `dirpath`, `dirnames` and `filenames` are identical to walk() output,
        and `dirfd` is a file descriptor referring to the directory `dirpath`.

        The advantage of fwalk() over walk() is that it's safe against symlink
        races (when follow_symlinks is False).

        If dir_fd is not None, it should be a file descriptor open to a directory,
          and top should be relative; top will then be relative to that directory.
          (dir_fd is always supported for fwalk.)

        Caution:
        Since fwalk() yields file descriptors, those are only valid until the
        next iteration step, so you should dup() them if you want to keep them
        for a longer period.

        Example:

        import os
        for root, dirs, files, rootfd in os.fwalk('python/Lib/email'):
            print(root, "consumes", end="")
            print(sum(os.stat(name, dir_fd=rootfd).st_size for name in files),
                  end="")
            print("bytes in", len(files), "non-directory files")
            if 'CVS' in dirs:
                dirs.remove('CVS')  # don't visit CVS directories
        """
        if not isinstance(top, int) or not hasattr(top, '__index__'):
            top = fspath(top)
        # Note: To guard against symlink races, we use the standard
        # lstat()/open()/fstat() trick.
        if not follow_symlinks:
            orig_st = stat(top, follow_symlinks=False, dir_fd=dir_fd)
        topfd = open(top, O_RDONLY, dir_fd=dir_fd)
        try:
            if (follow_symlinks or (st.S_ISDIR(orig_st.st_mode) and
                                    path.samestat(orig_st, stat(topfd)))):
                yield from _fwalk(topfd, top, isinstance(top, bytes),
                                  topdown, onerror, follow_symlinks)
        finally:
            close(topfd)

Notice it `close`s itself when the generator is completed. This is why the new implementation has a `for` loop with a `pass` body in `finally`

**Comment by c0llab0rat0r** [comment]
Transient windows build failure https://travis-ci.com/github/ipfs-shipyard/py-ipfs-http-client/jobs/497514288

File: tox.ini
Comment (ntninja) [reply tox.ini:24]: Please add it back. `py-cid` is used directly in the test (not just through `pytest-cid`), so there should be a direct dependency on it as well. (You can refrain from adding a specific version through, so that there will be no conflict between the one used by `pytest-cid` and the one specified directly.)

File: ipfshttpclient/filescanner.py
Comment (ntninja) [reply ipfshttpclient/filescanner.py:13]: Well frankly, I have never used PyCharm and I believe your overstating things. **But** if you say it's *that* useful, we can keep it if you add a comment on why it is done.

File: ipfshttpclient/filescanner.py
Comment (ntninja) [reply ipfshttpclient/filescanner.py]: Using `ty.Any` means that it will be compatible with functions that expect a `SendType` other than `None` (which it is), so I think its better to use `ty.Any` since it more accurately reflects what this type is and isn't compatible with.

File: ipfshttpclient/filescanner.py
Comment (ntninja) [reply ipfshttpclient/filescanner.py:693]: Every generator has `.close()`, `.send(…)` and `.throw(…)` methods implicitly. Calling `.close()` on the generator will raise a `GeneratorExit` exception at the `yield` that the generator is currently blocked on. Any `yield from`s will propagate their received exceptions and values “down-wards” into the generators *they* are blocked on. See [PEP-342](https://www.python.org/dev/peps/pep-0342/) for details on the “new” since Python 2.5 generators, that are more than simple iterables and [PEP-380](https://www.python.org/dev/peps/pep-0380/#formal-semantics) for the instructive definition of `yield from`.

File: ipfshttpclient/filescanner.py
Comment (ntninja) [reply ipfshttpclient/filescanner.py:662]: Why does it need this when we already have a `type: ignore` comment below?

If the answer is “because PyCharm doesn't understand that” then please remove this comment anyways. It is not unreasonable for PyCharm using people to just ignore the warning after looking at that line and I want to keep PyCharm-specific changes to an absolute minimum. Not everyone can and wants to use PyCharm and so they cannot use PyCharm's type checker and (like me) won't know what that comment is about, OTOH any Python dev can easily access mypy thanks to `tox` and the shipped config so that's what is primarily used here.

File: ipfshttpclient/filescanner_ty.py
Comment (ntninja) [reply ipfshttpclient/filescanner_ty.py:8]: Yes, but mypy *could* support it and there is no question about which syntax would be used to describe this. The mypy bug just is that it currently doesn't support it although it should. If other checkers don't support it by the time mypy gets around to fixing this, I'm sure they'll add support soon enough.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:614]: No, it doesn't type check with `os.fsencode`.

Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:614]: > do we really need a separate method here?

It helps solve the typing problems, and makes the code more readable.

The larger the methods get, the more the type linters (and humans) will struggle with them.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:662]: I'll remove it. The scope of this issue was a single line.

In some sense, it's hiding a problem that should be flagged anyways.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:693]: Thanks for the additional info. I'll take a closer look at this again.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py]: Got it, added a commit to use `Any` instead of `None`

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:693]: I've made an update that addresses this. It deduplicates the file descriptor closing (I found another identical block of code).

Out of curiosity, why is walk() a generator class instead of a generator function?

**Comment by c0llab0rat0r** [comment]
Of course the Windows builds failed again :sob: 
https://travis-ci.com/github/ipfs-shipyard/py-ipfs-http-client/jobs/497525637

@ntninja I think I've incorporated your feedback.

**Comment by c0llab0rat0r** [comment]
OMG a green build

:smile: :rocket: :100: :laughing: 

I'm buying a lottery ticket

File: ipfshttpclient/filescanner.py
Comment (ntninja) [reply ipfshttpclient/filescanner.py:693]: > I've made an update that addresses this. It deduplicates the file descriptor closing (I found another identical block of code).

Thank you! Never noticed that!

> Out of curiosity, why is walk() a generator class instead of a generator function?

You got me there for a second! :slightly_smiling_face: 

As far as I can tell, it's so that it can act as both a generator and a context-manager. So the following works:

```py
with filescanner.walk("./bla") as scanner:
    for item in scanner:
        process_item(item)  # Something that could raise
```

Instead of more awkward:

```py
scanner = filescanner.walk("./bla")
try:
    for item in scanner:
        process_item(item)  # Something that could raise
finally:
    scanner.close()
```

**Review by ntninja — APPROVED** [review]

**Comment by ntninja** [comment]
> OMG a green build
> 
> smile rocket 100 laughing
> 
> I'm buying a lottery ticket

I've upped the timeout on Windows builds to two minutes, if it still times out I'll xfail it.

File: ipfshttpclient/filescanner.py
Comment (c0llab0rat0r) [reply ipfshttpclient/filescanner.py:693]: The former is much nicer than the latter.

File: ipfshttpclient/filescanner_ty.py
Comment (ntninja) [reply ipfshttpclient/filescanner_ty.py]: (I just noticed you didn't actually do this, added it now.)

**Comment by ntninja** [comment]
Did you change anything? I just manually squashed your commits to make for a more readable git history, so it would have been a bad moment.

**Comment by c0llab0rat0r** [comment]
> Did you change anything? I just manually squashed your commits to make for a more readable git history, so it would have been a bad moment.

Nothing was changed; I rebased against master in case this complex change set would fail the build even if there are no git merge conflicts.

Incidentally it looks like the 3.6 build failed with a transient network error. I think we should have a green build before merging this change.

I'd like to keep the `AnyStr` commit separate, so the rest is easier to review without looking a bunch of `ty.AnyStr -> AnyStr` changes mixed in.

Is it fine with you if I squash all the commits except that one (which will also tickle and rerun the build)?

**Comment by c0llab0rat0r** [comment]
It looks like you merged these changes into master, but this PR is still open. 

Was that a mistake?

**Comment by c0llab0rat0r** [comment]
@ntninja I think you had an accident in `master` - you may need to rebase it to remove the commits from this PR and force push master.

**Comment by c0llab0rat0r** [comment]
It looks like the build in the last commit in master is green. I'm going to review what's different between master and this branch and see if this PR can simply be closed at this point.

**Comment by c0llab0rat0r** [comment]
I reviewed this branch and master and I believe the changes were merged into master correctly in spite of what may have been you doing the merge and me doing a rebase force push at the same time.

**Pull request closed without merging**
